File: clients/python/fever_client/models/search_request.py
from __future__ import annotations
import pprint
import re  # noqa: F401
import json

from pydantic import BaseModel, ConfigDict, Field, StrictBool, StrictFloat, StrictInt, StrictStr, field_validator
from typing import Any, ClassVar, Dict, List, Optional, Union
from typing import Optional, Set
from typing_extensions import Self
from pydantic_core import to_jsonable_python

class SearchRequest(BaseModel):
    after: Optional[StrictStr] = Field(default=None, description="EXIF capture time >= this ISO datetime.")
    before: Optional[StrictStr] = Field(default=None, description="EXIF capture time <= this ISO datetime.")
    collapse_duplicates: Optional[StrictBool] = Field(default=True, description="Collapse near-duplicate items so each result is a distinct representative; the folded-in duplicates are listed on each hit.")
    customer_id: Optional[StrictStr] = Field(default=None, description="Restrict results to media tagged with this customer_id.")
    device: Optional[StrictStr] = Field(default=None, description="EXIF camera make/model substring.")
    dupe_threshold: Optional[Union[StrictFloat, StrictInt]] = Field(default=None, description="Cosine similarity cutoff 0..1 for treating two items as duplicates; higher is stricter. Defaults to the appliance setting when omitted.")
    exclude_tags: Optional[List[StrictStr]] = Field(default=None, description="Exclude blobs carrying any of these tags (ingest tags and/or zero-shot autotags), e.g. ['people'] to build a no-people dataset.")
    image: Optional[StrictStr] = Field(default=None, description="Base64 image for image-to-image search / dedup.")
    image_blob_id: Optional[StrictInt] = Field(default=None, description="Search with the vector of an already-indexed image (no re-upload, no re-embed).")
    include_tags: Optional[List[StrictStr]] = Field(default=None, description="Only blobs carrying at least one of these tags (ANY semantics: e.g. a list of geo:<place> tags an integrator's UI offers). Combine with exclude_tags as you like.")
    indexed_res: Optional[StrictInt] = Field(default=None, description="Restrict results to blobs indexed at exactly this resolution (pixels; see IngestRequest.target_res). Blobs ingested before this filter existed have no recorded resolution and are excluded whenever this is set.")
    k: Optional[StrictInt] = Field(default=10, description="Max results.")
    kind: Optional[StrictStr] = Field(default=None, description="Constrain results to one media kind; unset scans every kind. Video hits attach seek_ms; audio/transcript hits attach span.")
    max_height: Optional[StrictInt] = Field(default=None, description="Restrict results to blobs at most this many pixels tall (stored media dimensions; blobs with unknown dimensions are excluded whenever this is set).")
    max_width: Optional[StrictInt] = Field(default=None, description="Restrict results to blobs at most this many pixels wide (stored media dimensions; blobs with unknown dimensions are excluded whenever this is set).")
    min_score: Optional[Union[StrictFloat, StrictInt]] = Field(default=0.0, description="Confidence cutoff 0..1; for text this is the reranker's match probability.")
    near: Optional[List[Union[StrictFloat, StrictInt]]] = Field(default=None, description="Geo filter [lat, lon, radius_km] from EXIF GPS.")
    sort: Optional[StrictStr] = Field(default=None, description="'relevance' (default) is the fused/calibrated ordering; 'distance' reorders results nearest-first and requires the near filter - e.g. an integrator's map view. An explicit ordering knob, not another relevance signal.")
    text: Optional[StrictStr] = Field(default=None, description="Text query (meaning-based, not keyword).")
    __properties: ClassVar[List[str]] = ["after", "before", "collapse_duplicates", "customer_id", "device", "dupe_threshold", "exclude_tags", "image", "image_blob_id", "include_tags", "indexed_res", "k", "kind", "max_height", "max_width", "min_score", "near", "sort", "text"]

    @field_validator('kind')
    def kind_validate_enum(cls, value):
        if value is None:
            return value

        if value not in set(['image', 'video', 'audio', 'pdf', 'doc', 'note']):
            raise ValueError("must be one of enum values ('image', 'video', 'audio', 'pdf', 'doc', 'note')")
        return value

    @field_validator('sort')
    def sort_validate_enum(cls, value):
        if value is None:
            return value

        if value not in set(['relevance', 'distance']):
            raise ValueError("must be one of enum values ('relevance', 'distance')")
        return value

    model_config = ConfigDict(
        validate_by_name=True,
        validate_by_alias=True,
        validate_assignment=True,
        protected_namespaces=(),
    )

    def to_str(self) -> str:
        return pprint.pformat(self.model_dump(by_alias=True))

    def to_json(self) -> str:
        return json.dumps(to_jsonable_python(self.to_dict()))

    @classmethod
    def from_json(cls, json_str: str) -> Optional[Self]:
        return cls.from_dict(json.loads(json_str))

    def to_dict(self) -> Dict[str, Any]:
        excluded_fields: Set[str] = set([
        ])

        _dict = self.model_dump(
            by_alias=True,
            exclude=excluded_fields,
            exclude_none=True,
        )
        if self.after is None and "after" in self.model_fields_set:
            _dict['after'] = None

        if self.before is None and "before" in self.model_fields_set:
            _dict['before'] = None

        if self.customer_id is None and "customer_id" in self.model_fields_set:
            _dict['customer_id'] = None

        if self.device is None and "device" in self.model_fields_set:
            _dict['device'] = None

        if self.dupe_threshold is None and "dupe_threshold" in self.model_fields_set:
            _dict['dupe_threshold'] = None

        if self.exclude_tags is None and "exclude_tags" in self.model_fields_set:
            _dict['exclude_tags'] = None

        if self.image is None and "image" in self.model_fields_set:
            _dict['image'] = None

        if self.image_blob_id is None and "image_blob_id" in self.model_fields_set:
            _dict['image_blob_id'] = None

        if self.include_tags is None and "include_tags" in self.model_fields_set:
            _dict['include_tags'] = None

        if self.indexed_res is None and "indexed_res" in self.model_fields_set:
            _dict['indexed_res'] = None

        if self.kind is None and "kind" in self.model_fields_set:
            _dict['kind'] = None

        if self.max_height is None and "max_height" in self.model_fields_set:
            _dict['max_height'] = None

        if self.max_width is None and "max_width" in self.model_fields_set:
            _dict['max_width'] = None

        if self.near is None and "near" in self.model_fields_set:
            _dict['near'] = None

        if self.sort is None and "sort" in self.model_fields_set:
            _dict['sort'] = None

        if self.text is None and "text" in self.model_fields_set:
            _dict['text'] = None

        return _dict

    @classmethod
    def from_dict(cls, obj: Optional[Dict[str, Any]]) -> Optional[Self]:
        if obj is None:
            return None

        if not isinstance(obj, dict):
            return cls.model_validate(obj)

        _obj = cls.model_validate({
            "after": obj.get("after"),
            "before": obj.get("before"),
            "collapse_duplicates": obj.get("collapse_duplicates") if obj.get("collapse_duplicates") is not None else True,
            "customer_id": obj.get("customer_id"),
            "device": obj.get("device"),
            "dupe_threshold": obj.get("dupe_threshold"),
            "exclude_tags": obj.get("exclude_tags"),
            "image": obj.get("image"),
            "image_blob_id": obj.get("image_blob_id"),
            "include_tags": obj.get("include_tags"),
            "indexed_res": obj.get("indexed_res"),
            "k": obj.get("k") if obj.get("k") is not None else 10,
            "kind": obj.get("kind"),
            "max_height": obj.get("max_height"),
            "max_width": obj.get("max_width"),
            "min_score": obj.get("min_score") if obj.get("min_score") is not None else 0.0,
            "near": obj.get("near"),
            "sort": obj.get("sort"),
            "text": obj.get("text")
        })
        return _obj

File: clients/csharp/src/LowdownLabs.Fever/Model/SearchRequest.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.IO;
using System.Runtime.Serialization;
using System.Text;
using System.Text.RegularExpressions;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using Newtonsoft.Json.Linq;
using System.ComponentModel.DataAnnotations;
using FileParameter = LowdownLabs.Fever.Client.FileParameter;
using OpenAPIDateConverter = LowdownLabs.Fever.Client.OpenAPIDateConverter;

namespace LowdownLabs.Fever.Model
{
    [DataContract(Name = "SearchRequest")]
    public partial class SearchRequest : IValidatableObject
    {
        [JsonConverter(typeof(StringEnumConverter))]
        public enum KindEnum
        {
            [EnumMember(Value = "image")]
            Image = 1,

            [EnumMember(Value = "video")]
            Video = 2,

            [EnumMember(Value = "audio")]
            Audio = 3,

            [EnumMember(Value = "pdf")]
            Pdf = 4,

            [EnumMember(Value = "doc")]
            Doc = 5,

            [EnumMember(Value = "note")]
            Note = 6
        }

        [DataMember(Name = "kind", EmitDefaultValue = true)]
        public KindEnum? Kind { get; set; }
        [JsonConverter(typeof(StringEnumConverter))]
        public enum SortEnum
        {
            [EnumMember(Value = "relevance")]
            Relevance = 1,

            [EnumMember(Value = "distance")]
            Distance = 2
        }

        [DataMember(Name = "sort", EmitDefaultValue = true)]
        public SortEnum? Sort { get; set; }
        public SearchRequest(string after = default, string before = default, bool collapseDuplicates = true, string customerId = default, string device = default, decimal? dupeThreshold = default, List<string> excludeTags = default, string image = default, int? imageBlobId = default, List<string> includeTags = default, int? indexedRes = default, int k = 10, KindEnum? kind = default, int? maxHeight = default, int? maxWidth = default, decimal minScore = 0.0M, List<decimal> near = default, SortEnum? sort = default, string text = default)
        {
            this.After = after;
            this.Before = before;
            this.CollapseDuplicates = collapseDuplicates;
            this.CustomerId = customerId;
            this.Device = device;
            this.DupeThreshold = dupeThreshold;
            this.ExcludeTags = excludeTags;
            this.Image = image;
            this.ImageBlobId = imageBlobId;
            this.IncludeTags = includeTags;
            this.IndexedRes = indexedRes;
            this.K = k;
            this.Kind = kind;
            this.MaxHeight = maxHeight;
            this.MaxWidth = maxWidth;
            this.MinScore = minScore;
            this.Near = near;
            this.Sort = sort;
            this.Text = text;
        }

        [DataMember(Name = "after", EmitDefaultValue = true)]
        public string After { get; set; }

        [DataMember(Name = "before", EmitDefaultValue = true)]
        public string Before { get; set; }

        [DataMember(Name = "collapse_duplicates", EmitDefaultValue = true)]
        public bool CollapseDuplicates { get; set; }

        [DataMember(Name = "customer_id", EmitDefaultValue = true)]
        public string CustomerId { get; set; }

        [DataMember(Name = "device", EmitDefaultValue = true)]
        public string Device { get; set; }

        [DataMember(Name = "dupe_threshold", EmitDefaultValue = true)]
        public decimal? DupeThreshold { get; set; }

        [DataMember(Name = "exclude_tags", EmitDefaultValue = true)]
        public List<string> ExcludeTags { get; set; }

        [DataMember(Name = "image", EmitDefaultValue = true)]
        public string Image { get; set; }

        [DataMember(Name = "image_blob_id", EmitDefaultValue = true)]
        public int? ImageBlobId { get; set; }

        [DataMember(Name = "include_tags", EmitDefaultValue = true)]
        public List<string> IncludeTags { get; set; }

        [DataMember(Name = "indexed_res", EmitDefaultValue = true)]
        public int? IndexedRes { get; set; }

        [DataMember(Name = "k", EmitDefaultValue = false)]
        public int K { get; set; }

        [DataMember(Name = "max_height", EmitDefaultValue = true)]
        public int? MaxHeight { get; set; }

        [DataMember(Name = "max_width", EmitDefaultValue = true)]
        public int? MaxWidth { get; set; }

        [DataMember(Name = "min_score", EmitDefaultValue = false)]
        public decimal MinScore { get; set; }

        [DataMember(Name = "near", EmitDefaultValue = true)]
        public List<decimal> Near { get; set; }

        [DataMember(Name = "text", EmitDefaultValue = true)]
        public string Text { get; set; }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            sb.Append("class SearchRequest {\n");
            sb.Append("  After: ").Append(After).Append("\n");
            sb.Append("  Before: ").Append(Before).Append("\n");
            sb.Append("  CollapseDuplicates: ").Append(CollapseDuplicates).Append("\n");
            sb.Append("  CustomerId: ").Append(CustomerId).Append("\n");
            sb.Append("  Device: ").Append(Device).Append("\n");
            sb.Append("  DupeThreshold: ").Append(DupeThreshold).Append("\n");
            sb.Append("  ExcludeTags: ").Append(ExcludeTags).Append("\n");
            sb.Append("  Image: ").Append(Image).Append("\n");
            sb.Append("  ImageBlobId: ").Append(ImageBlobId).Append("\n");
            sb.Append("  IncludeTags: ").Append(IncludeTags).Append("\n");
            sb.Append("  IndexedRes: ").Append(IndexedRes).Append("\n");
            sb.Append("  K: ").Append(K).Append("\n");
            sb.Append("  Kind: ").Append(Kind).Append("\n");
            sb.Append("  MaxHeight: ").Append(MaxHeight).Append("\n");
            sb.Append("  MaxWidth: ").Append(MaxWidth).Append("\n");
            sb.Append("  MinScore: ").Append(MinScore).Append("\n");
            sb.Append("  Near: ").Append(Near).Append("\n");
            sb.Append("  Sort: ").Append(Sort).Append("\n");
            sb.Append("  Text: ").Append(Text).Append("\n");
            sb.Append("}\n");
            return sb.ToString();
        }

        public virtual string ToJson()
        {
            return Newtonsoft.Json.JsonConvert.SerializeObject(this, Newtonsoft.Json.Formatting.Indented);
        }

        IEnumerable<ValidationResult> IValidatableObject.Validate(ValidationContext validationContext)
        {
            yield break;
        }
    }

}
